backend/test_observer/data_access/repository.py

```diff
@@ -130,6 +139,35 @@ def get_artefacts_by_family(
     return query.all()
 
 
+def get_historic_test_executions(
+    session: Session,
+    test_execution: TestExecution,
+) -> list[TestExecution]:
+    current_artefact = test_execution.artefact_build.artefact
+    return (
+        session.query(TestExecution)
+        .join(TestExecution.artefact_build)
+        .join(ArtefactBuild.artefact)
+        .options(
+            contains_eager(TestExecution.artefact_build).contains_eager(
+                ArtefactBuild.artefact
+            )
+        )
+        .filter(
+            Artefact.name == current_artefact.name,
+            Artefact.store == current_artefact.store,
+            Artefact.repo == current_artefact.repo,
+            Artefact.series == current_artefact.series,
+            TestExecution.environment_id == test_execution.environment_id,
+            TestExecution.id < test_execution.id,
+        )
+        .options(joinedload(TestExecution.test_results, innerjoin=True))
+        .order_by(TestExecution.created_at.desc())
+        .limit(HISTORIC_TEST_RESULT_COUNT)
+        .all()
+    )
```

Comment:
One suggestion expand on this SQL query to retrieve the test results grouped by test case and sorted already. This is likely more efficient

Reply:
I have now rewritten this. The only issue I couldn't resolve is how to generate a tuple of (version, status) on SQL-level, so I return two separate arrays, one for status, one for version and parse them into HistoricTestResult object in Python. Do you have any suggestions on how to do this better?

Reply:
After investigating this further I found that this query is quite complex. Looking at your latest code I think (not sure) that it has issues. Perhaps we should simplify this into separate problems and it should be ok to make use of python (I take back doing everything in sql land, at least group by is much easier on python land). It would be great to test all scenarios. But that will be challenging if they are all lumped into one function. So having smaller functions each with one purpose (even if each function is a sql subquery) would be much easier to test.

In high level we need the following:
1. Latest artefact builds per artefact and architecture. This would be a sql stmt and would be tested alone. We may already have code for this. Although I recall it uses group by. According to [this stackoverflow answer](https://stackoverflow.com/a/1313293), it's much faster to use self join over group by.
2. Latest test executions per artefact build and environment. Again this would be a sql stmt and would be tested alone. The solution will probably be very similar to the one above.
3. One big sql join that brings artefacts, latest artefact builds, latest test executions, environments, test results and test cases together. We could also sort by artefact id here.
4. One big python group by. Python's group by would create an iterable of grouped items which is what we need. This group by would need to use artefact id, artefact track, artefact series, artefact repo (not sure if group by can handle nulls the way we want, we may have to separate them into two depending on family), environment, test case.

Might have missed something, and there might be further room for improvement. But most importantly we need to test this, and testing individual properly parts is much easier than testing the whole properly

Reply:
You're absolutely right, maybe I worded it wrong. By artefact builds per architecture and artefact, I meant that the architecure is taken into consideration as to avoid different merging different architectures. But we still need to only get the latest build for each architecture. Because each architecure could have multiple builds. The same is for test executions, although now that I think more about it, this may not be necessary for test executions. Since we have a unique constraint on artefact build id and environment id. This means we can't have two test executions for the same environment and build (it just overrides the previous one). Therefore it's probably safe to skip the point about latest test executions

Reply:
I see, you are right that if we have a list of artefact builds, for each of them there will be at most 1 Test Execution we need to consider.

So I guess your suggestion is to find the historic test results in a top-to-bottom style:
1. Find all matching artefact builds, given Artefact identifiers (artefact name, track, series, repo) and Architecture
2. With the matched Artefact Builds from (1), we find the related TestExecutions and filter based on Environment ID
3. Find all the test results associated with the latest 10 test executions from (2)
4. Parse and group them on Python level

Reply:
Sounds about right, hopefully...

Actually, this begs the question. If we don't allow multiple test executions of the same environment and build, why do we allow multiple builds of the same architecture and artefacts.

Reply:
Hmm, I think according to the way we start test executions, there shouldn't be duplicated artefact builds. If I'm not wrong the revision count is sort-of identifier after the artefact_id. Although we don't have a constraint for this on the db-level. I wonder if we have a case of duplicated artefact builds with the same revision number.

Reply:
Hmmm, you're right, we actually have a [unique index](https://github.com/canonical/test_observer/blob/main/backend/test_observer/data_access/models.py#L172-L190) on artefact_id, architecture and revision (if it existed). So that would mean we can only have one artefact build per artefact and architecture combination. That simplifies things, even in other parts of the code (we were trying to only get latest builds :sweat_smile:)

Reply:
Oh wait nvm I mean different revisions should be merged or rather only the last revision should be taken into consideration

Reply:
@omar-selo could you please take a look at whether my implementation now fits into what you were suggesting? I can add more tests for different edge-cases if needed.